staging/src/k8s.io/csi-translation-lib/translate_test.go

```diff
@@ -77,6 +79,117 @@ func TestTranslationStability(t *testing.T) {
 	}
 }
 
+func TestZoneTranslation(t *testing.T) {
+	testCases := []struct {
+		name        string
+		pv          *v1.PersistentVolume
+		topologyKey string
+	}{
+		{
+			name: "GCE PD PV Source",
+			pv: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						v1.LabelZoneFailureDomain: "us-east-1",
+						v1.LabelZoneRegion:        "us-east-1a",
+					},
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+							PDName:    "test-disk",
+							FSType:    "ext4",
+							Partition: 0,
+							ReadOnly:  false,
+						},
+					},
+				},
+			},
+			topologyKey: plugins.GCEPDTopologyKey,
+		},
+		{
+			name: "AWS EBS PV Source",
+			pv: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						v1.LabelZoneFailureDomain: "us-east-1",
+						v1.LabelZoneRegion:        "us-east-1a",
+					},
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						AWSElasticBlockStore: &v1.AWSElasticBlockStoreVolumeSource{
+							VolumeID:  "vol01",
+							FSType:    "ext3",
+							Partition: 1,
+							ReadOnly:  true,
+						},
+					},
+				},
+			},
+			topologyKey: plugins.AWSEBSTopologyKey,
+		},
+		{
+			name: "Cinder PV Source",
+			pv: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						v1.LabelZoneFailureDomain: "us-east-1",
+						v1.LabelZoneRegion:        "us-east-1a",
+					},
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						Cinder: &v1.CinderPersistentVolumeSource{
+							VolumeID: "vol1",
+							FSType:   "ext4",
+							ReadOnly: false,
+						},
+					},
+				},
+			},
+			topologyKey: plugins.CinderTopologyKey,
+		},
+	}
+	for _, test := range testCases {
+		ctl := New()
+		t.Logf("Testing %v", test.name)
+
+		zone := test.pv.ObjectMeta.Labels[v1.LabelZoneFailureDomain]
+
+		// Translate to CSI PV and check translated node affinity
+		newCSIPV, err := ctl.TranslateInTreePVToCSI(test.pv)
+		if err != nil {
+			t.Errorf("Error when translating to CSI: %v", err)
+		}
+
+		if !isNodeAffinitySet(newCSIPV, test.topologyKey, zone) {
```

Comment:
I'll change this to cover cases where the topology is already set.

Reply:
Can we also add the expected NodeAffinity afterwards?  ie, zonal pd and regional pd are different.

Reply:
Expanded unit tests